libs/mrwebrtc/src/media/device_video_track_source.cpp

```diff
@@ -252,29 +296,449 @@ ErrorOr<RefPtr<DeviceVideoTrackSource>> DeviceVideoTrackSource::Create(
+      return;
+    }
+
+    // Get its supported capture formats
+    auto captureFormatList = vcd->getSupportedFormats();
+    for (auto&& captureFormat : *captureFormatList) {
+      mrsVideoCaptureFormatInfo format_info{};
+      format_info.width = captureFormat->get_width();
+      format_info.height = captureFormat->get_height();
+      format_info.framerate = captureFormat->get_framerateFloat();
+      format_info.fourcc = captureFormat->get_fourcc();
+      // When VideoEncodingProperties.Subtype() contains a GUID, the
+      // conversion to FOURCC fails and returns FOURCC_ANY. So ignore
+      // those formats, as we don't know their encoding.
+      if (format_info.fourcc != libyuv::FOURCC_ANY) {
+        enumerator.yield(&format_info);
+      }
+    }
+  });
+#else
+  std::unique_ptr<webrtc::VideoCaptureModule::DeviceInfo> info(
+      webrtc::VideoCaptureFactory::CreateDeviceInfo());
+  if (!info) {
+    return Error(Result::kUnknownError);
+  }
+  Enumerator<const mrsVideoCaptureFormatInfo*, mrsResult> enumerator(
+      enum_callback, end_callback, mrsResult::kSuccess);
+  int num_devices = info->NumberOfDevices();
+  for (int device_idx = 0; device_idx < num_devices; ++device_idx) {
+    // Filter devices by name
+    constexpr uint32_t kSize = 256;
+    char name[kSize] = {0};
+    char id[kSize] = {0};
+    if (info->GetDeviceName(device_idx, name, kSize, id, kSize) == -1) {
+      continue;
+    }
+    if (device_id != id) {
+      continue;
+    }
+
+    // Enum video capture formats
+    int32_t num_capabilities = info->NumberOfCapabilities(id);
+    for (int32_t cap_idx = 0; cap_idx < num_capabilities; ++cap_idx) {
+      webrtc::VideoCaptureCapability capability{};
+      if (info->GetCapability(id, cap_idx, capability) != -1) {
+        mrsVideoCaptureFormatInfo format_info{};
+        format_info.width = capability.width;
+        format_info.height = capability.height;
+        format_info.framerate = (float)capability.maxFPS;
+        format_info.fourcc = FourCCFromVideoType(capability.videoType);
+        // Ignore unknown capture formats
+        if (format_info.fourcc != libyuv::FOURCC_ANY) {
+          enumerator.yield(&format_info);
+        }
+      }
+    }
+
+    break;
+  }
+#endif
+  return Error::None();
+}
+
 DeviceVideoTrackSource::DeviceVideoTrackSource(
     RefPtr<GlobalFactory> global_factory,
-    rtc::scoped_refptr<webrtc::VideoTrackSourceInterface> source) noexcept
+    rtc::scoped_refptr<webrtc::VideoTrackSourceInterface> source
+#if defined(MR_SHARING_ANDROID)
+    ,
+    jobject java_video_capturer
+#endif  // defined(MR_SHARING_ANDROID)
+    ) noexcept
     : VideoTrackSource(std::move(global_factory),
                        ObjectType::kDeviceVideoTrackSource,
-                       std::move(source)) {}
+                       std::move(source))
+#if defined(MR_SHARING_ANDROID)
+      ,
+      java_video_capturer_(java_video_capturer)
+#endif  // defined(MR_SHARING_ANDROID)
+{
+}
```

Comment:
nit: it would probably be nicer to have the Android logic in a subclass of DeviceTrackSource (or at least to #ifdef the whole constructor here)

Reply:
Separated the #ifdef for clarity.